### PR TITLE
Pygrams2

### PR DESCRIPTION
**Pull request by thanasions**

### CODE REVIEW

File: scripts/data_factory.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import pandas as pd
+
+
+class DataFactory(object):
```
Comment (IanGrimstead) [inline scripts/data_factory.py]: Strip out class, just have the function - no need for class overhead

File: scripts/output_factory.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class OutputFactory(object):
+
+    @staticmethod
```
Comment (IanGrimstead) [inline scripts/output_factory.py]: Remove class overhead...

File: scripts/tfidf_wraper.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import numpy as np
```
Comment (IanGrimstead) [inline scripts/tfidf_wraper.py]: Rename to `tfidf_wrapper.py`

File: scripts/output_factory.py [hunk 4]
```diff
@@ -1,14 +1,40 @@
...
\ No newline at end of file
+import bz2
+import os
+import pickle
+
+from scripts.terms_graph import TermsGraph
+from scripts.visualization.wordclouds.multicloudplot import MultiCloudPlot
+
+
+def get(output_type, output, wordcloud_title=None, tfidf_reduce_obj=None, name=None, nterms=50, term_counts_mat=None):
```
Comment (IanGrimstead) [inline scripts/output_factory.py]: `get` name doesn't quite match what it does - as it doesn't return anything. `produce_output` instead? `generate_output` ?

File: scripts/tfidf_mask.py [hunk 5]
```diff
@@ -1,6 +1,108 @@
...
+        if unigrams:
+            self.__clean_unigrams(self.__max_bigram())
+
+        # normalize rows to text length
+        if norm_rows:
+            if doc_weights is None:
+                self.__doc_weights=np.ones(len(self.__text_abstracts))
+            self.__normalize_rows()
+
+        self.__unbias_ngrams(max_ngram_length)
```
Comment (IanGrimstead) [inline scripts/tfidf_mask.py]: unbias unigrams - feels like part of the pipeline rather then in here - ticket to split this out? Same for clean unigrams. The Mask should be just that, a means of removing (ignoring) ngrams?

File: scripts/output_factory.py (hunk 4 above)
Comment (thanasions) [reply scripts/output_factory.py]: good point.

File: scripts/tfidf_mask.py (hunk 5 above)
Comment (thanasions) [reply scripts/tfidf_mask.py]: yes, let's do tickets for those after we plugged in the tests.

File: pygrams2.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
+import argparse
+import os
+import sys
+
+from scripts.pipeline import Pipeline
+from scripts.utils.argschecker import ArgsChecker
+
+
+# -fc="Communications,Leadership, IT systems"
```
Comment (IanGrimstead) [inline pygrams2.py]: Remove commented out params? Move to readme / manual?

File: pygrams2.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    parser.add_argument("-ndl", "--normalize_doc_length", default=False, action="store_true",
+                        help="normalize tf-idf scores by document length")
+    parser.add_argument("-t", "--time", default=False, action="store_true", help="weight terms by time")
+
+    parser.add_argument("-pt", "--path", default='data', help="the data path")
+    parser.add_argument("-ih", "--id_header", default=None, help="the column name for the unique ID")
+    parser.add_argument("-th", "--text_header", default='abstract', help="the column name for the free text")
+    parser.add_argument("-dh", "--date_header", default=None, help="the column name for the date")
+    parser.add_argument("-fc", "--filter_columns", default=None, help="list of columns to filter by")
+    parser.add_argument("-fb", "--filter_by", default='union', choices=['union', 'intersection'],
```
Comment (IanGrimstead) [inline pygrams2.py]: Help doesn't make sense relative to options

File: pygrams2.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                        help="options are <all> <any> defaults to any. Returns filter where all are 'Yes' "
+                             "or any are 'Yes")
+
+    parser.add_argument("-p", "--pick", default='sum', choices=['median', 'max', 'sum', 'avg'],
+                        help="options are <median> <max> <sum> <avg>  defaults to sum. Average is over non zero values")
```
Comment (IanGrimstead) [inline pygrams2.py]: average (avg) and median...

File: pygrams2.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    parser.add_argument("-mt", "--month_to", default=None, help="The last month for the document cohort in MM format")
+
+    parser.add_argument("-np", "--num_ngrams_report", type=int, default=250,
+                        help="number of ngrams to return for report")
+    parser.add_argument("-nd", "--num_ngrams_wordcloud", type=int, default=250,
+                        help="number of ngrams to return for wordcloud")
+    parser.add_argument("-nf", "--num_ngrams_fdg", type=int, default=50,
+                        help="number of ngrams to return for fdg graph")
+
+    parser.add_argument("-ds", "--doc_source", default='USPTO-random-1000.pkl.bz2', help="the doc source to process")
```
Comment (IanGrimstead) [inline pygrams2.py]:
```suggestion
    parser.add_argument("-ds", "--doc_source", default='USPTO-random-1000.pkl.bz2', help="the document source to process")
```

File: pygrams2.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    parser.add_argument("-fs", "--focus_source", default='USPTO-random-1000.pkl.bz2',
+                        help="the doc source for the focus function")
```
Comment (IanGrimstead) [inline pygrams2.py]:
```suggestion
                        help="the document source for the focus function")
```

File: pygrams2.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+    parser.add_argument("-mn", "--min_n", type=int, choices=[1, 2, 3], default=1, help="the minimum ngram value")
+    parser.add_argument("-mx", "--max_n", type=int, choices=[1, 2, 3], default=3, help="the maximum ngram value")
+    parser.add_argument("-mdf", "--max_document_frequency", type=float, default=0.05,
+                        help="the maximum document frequency to contribute to TF/IDF")
+
+    parser.add_argument("-on", "--outputs_name", default='out', help="outputs filename")
+
+    parser.add_argument("-wt", "--wordcloud_title", default='tech terms', help="wordcloud title")
```
Comment (IanGrimstead) [inline pygrams2.py]:
```suggestion
    parser.add_argument("-wt", "--wordcloud_title", default='Popular terms', help="wordcloud title")
```

File: scripts/algorithms/tfidf.py [hunk 12]
```diff
@@ -178,6 +178,7 @@ def __init__(self, docs_df, ngram_range=(1, 3), max_document_frequency=0.3, toke
         self.__uni_factor = uni_factor
 
         num_docs_before_sift = self.__dataframe.shape[0]
+        print(self.__text_header)
```
Comment (IanGrimstead) [inline scripts/algorithms/tfidf.py]: `print` still required?

File: scripts/filter_terms.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            print('Loading model: '+ file_name)
+            self.__model = KeyedVectors.load_word2vec_format(self.__file_name)
+            self.__ngram_weights_vec = self.__get_embeddings_vec(threshold)
+
+
+    @property
+    def ngram_weights_vec(self):
+        return self.__ngram_weights_vec
+
+    # no need for both threshold and binary as threshold only needed in binary :)
```
Comment (IanGrimstead) [inline scripts/filter_terms.py]: Anything we need to attend to now?

File: scripts/output_factory.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    if output_type == 'report':
+        filename_and_path = os.path.join('outputs', 'reports', name + '.txt')
+        with open(filename_and_path, 'w') as file:
+            counter = 1
+            for score, term in output:
+                if counter < nterms:
+                    file.write(f' {term:30} {score:f}\n')
+                    print(f'{counter}. {term:30} {score:f}')
+                    counter += 1
+        print(output_type)
```
Comment (IanGrimstead) [inline scripts/output_factory.py]: No longer need `print`?
```suggestion
```

File: scripts/output_factory.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    elif output_type == 'graph':
+        print(output_type)
```
Comment (IanGrimstead) [inline scripts/output_factory.py]: No longer need `print`?
```suggestion
```

File: scripts/pipeline.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class Pipeline(object):
+    def __init__(self, data_filename, docs_mask_dict,  pick_method='sum', ngram_range=(1,3),
+                 normalize_rows=False, text_header='abstract', term_counts=False,
+                 pickled_tf_idf=False, max_df=0.1):
+
+        # load data
+        df = datafactory.get(data_filename)
+        user_ngrams = 'agriculture' # 'green, energy, wind, solar, engine'
```
Comment (IanGrimstead) [inline scripts/pipeline.py]: Should `user_ngrams` be hard-wired?

File: scripts/tfidf_mask.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    def __init__(self, tfidf_obj, ngram_range=(2,3), uni_factor=0.8):
+        self.__tfidf_matrix = tfidf_obj.tfidf_matrix
+        self.__feature_names = tfidf_obj.feature_names
+        self.__tfidf_mask = self.__tfidf_matrix.copy()
+        self.__tfidf_mask.data = np.ones(len(self.__tfidf_matrix.data))
+        self.__vectorizer = tfidf_obj.vectorizer
+        self.__uni_factor = uni_factor
+        self.__idf = tfidf_obj.idf
+
+        # self.__ngram_counts = csr_matrix(self.__ngram_counts, dtype=np.float64, copy=True)
```
Comment (IanGrimstead) [inline scripts/tfidf_mask.py]: Remove comment?

File: scripts/tfidf_mask.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+                    ngram_minus_front = ' '.join(big_ngram_terms[1:])
+                    ngram_minus_back = ' '.join(big_ngram_terms[:len(big_ngram_terms) - 1])
+
+                    idx_ngram_minus_front = self.__vectorizer.vocabulary_.get(ngram_minus_front)
+                    idx_ngram_minus_back = self.__vectorizer.vocabulary_.get(ngram_minus_back)
+
+                    indices_slice = self.__tfidf_matrix.indices[start_idx_ptr:end_idx_ptr]
+                    ngram_counts = self.__tfidf_matrix.data[j]/self.__idf[col_idx]
+                    # if i==95:
```
Comment (IanGrimstead) [inline scripts/tfidf_mask.py]: Remove debug comments?

File: scripts/tfidf_reduce.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+                if np.isnan(pick_value):
+                    pick_value = 0
+
+                ngrams_scores_tuple.append((pick_value, ngram))
+
+        ngrams_scores_tuple.sort(key=lambda tup: -tup[0])
+        return ngrams_scores_tuple
+
+    # obsolete
+    def extract_ngrams_from_docs_set(self, docs_subset, pick_method, verbose=True):
```
Comment (IanGrimstead) [inline scripts/tfidf_reduce.py]: Can we remove this?

File: scripts/tfidf_wrapper.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
+from scripts.text_processing import StemTokenizer, lowercase_strip_accents_and_ownership, WordAnalyzer
+from sklearn.feature_extraction.text import CountVectorizer
+from sklearn.feature_extraction.text import TfidfTransformer
+
+
+# add pickle option
```
Comment (IanGrimstead) [inline scripts/tfidf_wrapper.py]: replace comment with ticket?

File: scripts/utils/utils.py [hunk 21]
```diff
@@ -1,15 +1,38 @@
...
+    min_x, max_x = min(X), max(X)
+    diff_x = (max_x - min_x)
+    std_x = (np.array(X) - min_x) / diff_x
+    x_scaled = std_x*(1-min_val) + min_val
+    return x_scaled if not return_list else list(x_scaled)
+
+
+def w2vify(filein, fileout):
+    glove_file = datapath(filein)
+    tmp_file = get_tmpfile(fileout)
```
Comment (IanGrimstead) [inline scripts/utils/utils.py]: When is tmpfile deleted?
